### PR TITLE
Common deep copy refactor

### PR DESCRIPTION
**Pull request by scorbajio**
This change refactors the `Common` class to deep copy parameters on calls to `copy`. It aims to address issue #3593 and #3572.

### CODE REVIEW

**Review by acolytec3 — DISMISSED** [review]
LGTM

File: packages/common/test/chains.spec.ts [hunk 1]
```diff
@@ -23,6 +24,33 @@ describe('[Common/Chains]: Initialization / Chain params', () => {
...
+      console.log('dbg101')
+      console.log(cCopy)
+      assert.equal(cCopy.chainName(), 'mainnet', 'should return original chain name')
+      assert.equal(cCopy.chainId(), BigInt(1), 'should return original chain Id')
+    }
+
+    await setCommon()
+    await resetCommon()
+  })
+
```
Comment (holgerd77) [inline packages/common/test/chains.spec.ts]: I am honestly very confused by this whole setup, so these tests trigger the issue but were later *removed*?? 😂

I will copy (cherry-pick) them over in a new PR and add the changes I initially intended.

Comment (holgerd77) [reply packages/common/test/chains.spec.ts]: (and see if tests are passing)

File: packages/common/src/common.ts [hunk 2]
```diff
@@ -69,7 +71,7 @@ export class Common {
         (this._chainParams.customHardforks && this._chainParams.customHardforks[hf.name]),
     ])
     this._hardfork = this.DEFAULT_HARDFORK
-    this._params = { ...(opts.params ?? {}) } // copy
+    this._params = opts.params ? JSON.parse(JSON.stringify(opts.params)) : {} // copy
```
Comment (holgerd77) [inline packages/common/src/common.ts]: Ok, so this is actually the single line change I initially intended, likely with the ones from `updateParams()` and `resetParams()` (good catch!), no changes intended for `copy()` whatsoever.

Will tests these lines in a new PR (maybe again, eventually you've been to this point already), then this should already be enough for this (or a replacing) PR!

(we also should not mix this with #3593 , it's absolutely not clear that these are related)

File: packages/common/src/common.ts [hunk 3]
```diff
@@ -58,6 +59,7 @@ export class Common {
   public events: EventEmitter
 
   constructor(opts: CommonOpts) {
+    this._opts = opts
     this.events = new EventEmitter()
 
     this._chainParams = opts.chain
```
Comment (holgerd77) [inline packages/common/src/common.ts]: Ah, but you are right, we need additional treatment for `chain`. But we should do here, then we can avoid doing any `copy()` changes.

Also (more important): if we do *not* here, then the initial `Mainnet` (or whatever chain) structure gets modified *in the first place*. That's bad and we should very much avoid (same for the `params` structs, I stumbled upon this kind of failure).

File: packages/common/test/chains.spec.ts [hunk 4]
```diff
@@ -23,6 +24,33 @@ describe('[Common/Chains]: Initialization / Chain params', () => {
     )
   })
 
+  it('Deep copied common object should have parameters that are independent of the original copy', async () => {
+    let chainConfig: ChainConfig
+    let c: Common
+    const setCommon = async () => {
+      chainConfig = Mainnet
```
Comment (holgerd77) [inline packages/common/test/chains.spec.ts]: Haha, so funny, I am pretty sure (or: could very much imagine) that this is where you turned wrong: so, by associating `chainConfig = Mainnet` here and *then* manipulating this further down the line with `      chainConfig.chainId = 2` this then also manipulates `Mainnet` itself.

And *that* causes another test below in line 55 to fail. 😂 And the error looks pretty much like a "still failing" message if one is not looking pretty closely. 🙂

**Comment by holgerd77** [comment]
Have replaced with #3613, will close here.

Ok, but this work was definitely not "wasted", since this discovered various additional necessary deep-copy cases + could re-use the test setup! 🙏

**Pull request closed without merging**